File: ui/windows/src/Clambhook.Windows/Core/DaemonSupervisor.cs
using System.Diagnostics;

namespace Clambhook.Windows.Core;

public enum DaemonState
{
    Stopped,
    Starting,
    Running,
    Stopping,
    Failed
}

public sealed class DaemonSupervisor
{
    private Process? _process;

    public event Action? StateChanged;

    public DaemonState State { get; private set; } = DaemonState.Stopped;
    public string Message { get; private set; } = "";
    public bool IsRunning => _process is { HasExited: false };
    public bool IsBusy => State is DaemonState.Starting or DaemonState.Stopping;
    public string StateLabel => State switch
    {
        DaemonState.Stopped => "Daemon stopped",
        DaemonState.Starting => "Daemon starting",
        DaemonState.Running => "Daemon running",
        DaemonState.Stopping => "Daemon stopping",
        DaemonState.Failed => "Daemon failed",
        _ => "Daemon"
    };

    public Task StartAsync(AppSettings settings, string token, string appBaseDirectory, CancellationToken cancellationToken = default)
    {
        if (IsRunning)
        {
            SetState(DaemonState.Running, "daemon already running");
            return Task.CompletedTask;
        }

        SetState(DaemonState.Starting, "");
        var executable = ResolveExecutablePath(settings, appBaseDirectory);
        if (executable is null)
        {
            SetState(DaemonState.Failed, "clambhook daemon executable was not found");
            throw new FileNotFoundException("clambhook daemon executable was not found");
        }

        var startInfo = new ProcessStartInfo
        {
            FileName = executable,
            UseShellExecute = false,
            CreateNoWindow = true,
            WorkingDirectory = Path.GetDirectoryName(executable) ?? appBaseDirectory
        };
        foreach (var argument in BuildArgumentList(settings, token))
        {
            startInfo.ArgumentList.Add(argument);
        }

        try
        {
            _process = Process.Start(startInfo);
            if (_process is null)
            {
                throw new InvalidOperationException("failed to launch clambhook daemon");
            }

            _process.EnableRaisingEvents = true;
            _process.Exited += Process_Exited;
            SetState(DaemonState.Running, "daemon launched");
        }
        catch (Exception error)
        {
            _process = null;
            SetState(DaemonState.Failed, error.Message);
            throw;
        }

        return Task.CompletedTask;
    }

    public Task StopAsync()
    {
        if (_process is { HasExited: false } process)
        {
            SetState(DaemonState.Stopping, "");
            process.Exited -= Process_Exited;
            process.Kill(entireProcessTree: true);
            process.WaitForExit(3000);
            process.Dispose();
        }

        _process = null;
        SetState(DaemonState.Stopped, "daemon stopped");
        return Task.CompletedTask;
    }

    public static string? ResolveExecutablePath(AppSettings settings, string appBaseDirectory)
    {
        var configured = settings.DaemonPath.Trim();
        if (!string.IsNullOrEmpty(configured) && File.Exists(configured))
        {
            return configured;
        }

        var bundled = Path.Combine(appBaseDirectory, "clambhook.exe");
        return File.Exists(bundled) ? bundled : null;
    }

    public static string BuildArguments(AppSettings settings, string token)
    {
        return string.Join(" ", BuildArgumentList(settings, token).Select(argument =>
            argument.StartsWith('-') ? argument : Quote(argument)));
    }

    public static IReadOnlyList<string> BuildArgumentList(AppSettings settings, string token)
    {
        var normalized = settings.Normalized();
        var args = new List<string> { "-api", ApiListenAddress(normalized.ApiEndpoint) };
        if (!string.IsNullOrWhiteSpace(token))
        {
            args.Add("-api-token");
            args.Add(token.Trim());
        }

        if (!string.IsNullOrWhiteSpace(normalized.ConfigPath))
        {
            args.Add("-config");
            args.Add(normalized.ConfigPath);
        }

        return args;
    }

    public static string ApiListenAddress(string endpoint)
    {
        var normalized = AppSettings.NormalizeEndpoint(endpoint);
        if (Uri.TryCreate(normalized, UriKind.Absolute, out var uri) && !string.IsNullOrWhiteSpace(uri.Host))
        {
            var host = uri.Host;
            if (host.Contains(':') && !host.StartsWith('['))
            {
                host = $"[{host}]";
            }

            return uri.IsDefaultPort ? host : $"{host}:{uri.Port}";
        }

        return normalized;
    }

    private static string Quote(string value)
    {
        var escaped = value.Replace("\"", "\\\"");
        return $"\"{escaped}\"";
    }

    private void Process_Exited(object? sender, EventArgs e)
    {
        if (sender is not Process exited || !ReferenceEquals(exited, _process))
        {
            return;
        }

        var exitCode = 0;
        try
        {
            exitCode = exited.ExitCode;
        }
        catch
        {
        }

        exited.Dispose();
        _process = null;
        if (State != DaemonState.Failed)
        {
            SetState(
                exitCode == 0 ? DaemonState.Stopped : DaemonState.Failed,
                exitCode == 0 ? "daemon exited" : $"daemon exited with code {exitCode}");
        }
    }

    private void SetState(DaemonState state, string message)
    {
        State = state;
        Message = message;
        StateChanged?.Invoke();
    }
}

public static class Formatters
{
    public static string FormatRate(double bytesPerSecond)
    {
        var units = new[] { "B/s", "KB/s", "MB/s", "GB/s" };
        var value = bytesPerSecond;
        var unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }

        return unit == 0 ? $"{(int)value} {units[unit]}" : $"{value:0.0} {units[unit]}";
    }

    public static string FormatBytes(ulong bytes)
    {
        var units = new[] { "B", "KB", "MB", "GB" };
        var value = (double)bytes;
        var unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }

        return unit == 0 ? $"{(ulong)value} {units[unit]}" : $"{value:0.0} {units[unit]}";
    }

    public static string FormatDurationNs(long ns)
    {
        if (ns <= 0)
        {
            return "--";
        }

        var seconds = ns / 1_000_000_000;
        if (seconds < 1)
        {
            return $"{ns / 1_000_000} ms";
        }

        return seconds < 60 ? $"{seconds} s" : $"{seconds / 60} min";
    }

    public static string ServerLocation(ServerPayload server)
    {
        var parts = new[] { server.Geo.City, server.Geo.Country }.Where(part => !string.IsNullOrWhiteSpace(part));
        var location = string.Join(", ", parts);
        return string.IsNullOrWhiteSpace(location) ? server.Address : location;
    }
}

File: ui/linux/src/main/kotlin/com/clambhook/linux/daemon/DaemonSupervisor.kt
package com.clambhook.linux.daemon

import com.clambhook.linux.settings.AppSettings
import com.clambhook.linux.settings.normalized
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

enum class DaemonState { STOPPED, STARTING, RUNNING, STOPPING, FAILED }

data class DaemonStatus(val state: DaemonState, val message: String = "") {
    fun stateLabel(): String = when (state) {
        DaemonState.STOPPED -> "Daemon stopped"
        DaemonState.STARTING -> "Daemon starting"
        DaemonState.RUNNING -> "Daemon running"
        DaemonState.STOPPING -> "Daemon stopping"
        DaemonState.FAILED -> "Daemon failed"
    }
    fun stateIsBusy(): Boolean = state == DaemonState.STARTING || state == DaemonState.STOPPING
}

class DaemonSupervisor {
    private var process: Process? = null
    private val _status = MutableStateFlow(DaemonStatus(DaemonState.STOPPED))
    val status: StateFlow<DaemonStatus> = _status.asStateFlow()

    val isRunning: Boolean get() = process != null && process!!.isAlive

    fun start(settings: AppSettings, token: String, appBaseDir: String, licensePath: String? = null) {
        if (isRunning) { transition(DaemonState.RUNNING); return }
        transition(DaemonState.STARTING)
        try {
            val executable = resolveExecutablePath(settings, appBaseDir)
                ?: throw IllegalStateException("clambhook daemon executable was not found")
            val argv = buildArgv(settings, token, licensePath).toMutableList().also { it.add(0, executable) }
            val builder = ProcessBuilder(argv)
            process = builder.start()
            transition(DaemonState.RUNNING)
            watchProcess(process!!)
        } catch (e: Exception) {
            process = null
            transition(DaemonState.FAILED, e.message ?: "failed")
            throw e
        }
    }

    fun stop() {
        process?.let { if (it.isAlive) { transition(DaemonState.STOPPING); it.destroyForcibly() } }
        process = null
        transition(DaemonState.STOPPED)
    }

    private fun watchProcess(watched: Process) {
        Thread {
            try { watched.waitFor() } catch (e: Exception) {}
            if (process == watched) {
                process = null
                if (_status.value.state != DaemonState.FAILED) transition(DaemonState.STOPPED)
            }
        }.apply { isDaemon = true; start() }
    }

    private fun transition(next: DaemonState, message: String = "") {
        _status.value = DaemonStatus(next, message)
    }

    companion object {
        fun defaultAppBaseDir(): String = try {
            Paths.get("/proc/self/exe").toRealPath().parent.toString()
        } catch (e: Exception) {
            System.getProperty("user.dir")
        }

        fun resolveExecutablePath(settings: AppSettings, appBaseDir: String, searchPath: Boolean = true): String? {
            val configured = settings.daemonPath.trim()
            if (configured.isNotEmpty() && File(configured).exists()) return configured
            if (searchPath) {
                val pathExec = System.getenv("PATH").split(File.pathSeparator).map { File(it, "clambhook") }
                    .firstOrNull { it.exists() }?.absolutePath
                if (pathExec != null && pathExec.isNotEmpty()) return pathExec
            }
            val bundled = File(appBaseDir, "clambhook")
            return if (bundled.exists()) bundled.absolutePath else null
        }

        fun buildArguments(settings: AppSettings, token: String, licensePath: String? = null): String {
            return buildArgv(settings, token, licensePath).joinToString(" ") { if (it.startsWith("-")) it else "\"${it.replace("\"", "\\\"")}\"" }
        }

        fun buildArgv(settings: AppSettings, token: String, licensePath: String? = null): List<String> {
            val normalized = settings.normalized()
            val args = mutableListOf("-api", apiListenAddress(normalized.apiEndpoint))
            val trimmedToken = token.trim()
            if (trimmedToken.isNotEmpty()) { args.add("-api-token"); args.add(trimmedToken) }
            if (normalized.configPath.isNotEmpty()) { args.add("-config"); args.add(normalized.configPath) }
            if (licensePath != null && licensePath.trim().isNotEmpty()) { args.add("-license"); args.add(licensePath.trim()) }
            return args
        }

        fun apiListenAddress(endpoint: String): String {
            return try {
                val uri = java.net.URI(endpoint)
                val host = uri.host ?: ""
                val port = uri.port
                val addressHost = if (host.contains(":") && !host.startsWith("[")) "[$host]" else host
                if (host.isNotEmpty()) {
                    if (port >= 0) "$addressHost:$port" else addressHost
                } else endpoint
            } catch (e: Exception) {
                endpoint
            }
        }
    }
}